semgrep-core/core/Caching.ml

```diff
@@ -0,0 +1,376 @@
+(*
+   Decorate a pattern and target ASTs to make the suitable for memoization
+   during matching.
+*)
+
+module MV = Metavariable
+
+open Printf
+open AST_generic
+
+let debug = false
+
+(* A set of metavariables. Access cost is O(log n). *)
+module Names = AST_generic.String_set
```

Comment:
You know, in practice it might be better to use a list ... there are very few metavariables usually
in patterns, and in the past I sometimes found that plain list can be faster than binary-trees set.